### PR TITLE
feat(user status): automate user status for events

### LINKED ISSUE #31177: Set user status to busy when own calendar contains busy event
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

## Is your feature request related to a problem? Please describe.
Availability status seems not to be synced with Nextcloud Calendar like Microsoft Outlook with Teams.

### Describe the solution you'd like
When a "busy" event is in progress in Nextcloud Agenda, set availability status in Talk as "busy" or "to not disturb" and eventually share informations about the event in progress as status message (in option).

### Describe alternatives you've considered**
Microsoft does the same with Outlook and Teams/Skype for Business. If a calendar event is in progress with "busy" status, Teams status is "busy" or "in a meeting". If an event is "out of office", Teams status is "out of office".

Nextcloud is an open source and very powerful alternative to GAFAM, so it would be very interesting, in my opinion.

## Work packages

- [ ] Status automation
- [ ] User documentation
- [ ] Admin documentation (if any)

### Follow up
- https://github.com/nextcloud/server/issues/40760

### PR DESCRIPTION
**Pull request by miaulalala**
* Resolves: https://github.com/nextcloud/server/issues/31177

## Summary

The user status currently only takes into account the availability settings. Automatically set a user status to free or busy depending on their calendar transparency, event status and availability settings combined.

This can be done with a free/busy request just like the Calendar app does for the free/busy modal.

## How to test:

THe easiest way to test is to create 5- minute calendar events and have some sort of VAVAILABLITY setting.

The decision tree is the following:

Status decision tree:

USER DEFINED STATUS
  - Persistent status such as DND, OFFLINE, INVISIBLE?
    - leave the status alone
  - Not persistent?
    - get Calendar status
      - calendar checks for events (BUSY) and availability (within the availability and no event = ONLINE, event = AWAY + in a meeting status, outside availability = BUSY + away status)
      - calendar reports identical to existing status (i.e. online == online)
        - [leave status alone](https://www.youtube.com/watch?v=uzIGFXVJpQY)
      - calendar status is different (i. e. user has status ONLINE but calendar reports AWAY or BUSY)
        - process the new calendar status - in a meeting if a event is happening, away if outside availability

NOT USER DEFINED
  - get calendar status
 - process the new calendar status - in a meeting if a event is happening, away if outside availability

## TODO

- [ ] tests

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] ~~Screenshots before/after for front-end changes~~
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] ~~[Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)~~

### CODE REVIEW

**Comment by miaulalala** [comment]
I haven't figured out how to set the colour dot to something else yet, but other stuff works.

**Comment by miaulalala** [comment]
Having https://github.com/nextcloud/calendar/issues/3193 would be good @ChristophWurst - shall I add it as a followup?


Context:
image

**Comment by nimishavijay** [comment]
> Would be nice if that could be a DND icon but in yellow maybe?

I love the idea of using yellow for a busy state! We will also have to indicate it with icon/shape in addition to color, because would be difficult to distinguish between the DND and busy state for people with colorblindness:

No color blindness
image
Green-blind 
image
Red-blind 
image

How about using an icon that is similar to DND -- a focus icon. iOS has a crescent moon (like the Away status in NC) but Android had a [target-looking icon in a previous version](https://beebom.com/android-10-focus-mode/)

We could use something reminiscent of a focus mode like that? What do you think? (icons used here are [radiobox-marked](https://fonts.google.com/icons?selected=Material%20Icons%3Aradio_button_checked%3A), [circle-slice-8](https://pictogrammers.com/library/mdi/icon/circle-slice-8/), [record-circle](https://pictogrammers.com/library/mdi/icon/record-circle/), and [circle-double](https://pictogrammers.com/library/mdi/icon/circle-double/))
image


My vote would go for 1 or 3, what do you think? also @jancborchardt @marcoambrosini @szaimen

**Comment by jancborchardt** [comment]
@miaulalala regarding which calendars to take into account, I would say starting out with "Personal" and then any event which you are participant of in other calendars would be best. The second part might need work in other areas though?

---

About the state:

Microsoft Teams uses a filled red circle – similar to do not disturb but also reminiscing "recording in progress".
To distinguish it from the "Available" green circle, we can add a checkmark to the green circle. So we could align with Microsoft Teams to use a de-facto standard. It would also include adjusting the "Away" state from a crescent moon into a clock.
Also note they have an interesting "Out of office" state on the bottom. :)

(There might be possible conflict with the desktop client sync icon which can also be a white checkmark in a green circle. However this is also a more general symbol used elsewhere like we will also use in our admin overview that will come with 28 as well.)

image

**Comment by miaulalala** [comment]
So - any decision on the icon @jancborchardt @nimishavijay ?

I think the status alignment with teams needs to be a separate ticket and is not within the scope of this feature.

**Comment by nimishavijay** [comment]
> I think the status alignment with teams needs to be a separate ticket and is not within the scope of this feature.

Would it not be possible to just change the icons for all the statuses?
New icons:
Available: [check-circle](https://fonts.google.com/icons?selected=Material%20Icons%3Acheck_circle%3A) in `--color-success`
Busy: [circle](https://fonts.google.com/icons?selected=Material%20Icons%3Acircle%3A) in `--color-error`
Away: [watch-later](https://fonts.google.com/icons?selected=Material%20Icons%3Awatch_later%3A) in `--color-warning`
DND: [do-not-disturb-on](https://fonts.google.com/icons?selected=Material%20Icons%3Ado_not_disturb_on%3A) in `--color-error`
Offline: [trip-origin](https://fonts.google.com/icons?selected=Material%20Icons%3Atrip_origin%3A) in `--color-text-maxcontrast`

**Comment by jancborchardt** [comment]
Agree with @nimishavijay. If that is out of scope, a good first step would be to just do the same that we do when in a call – which is not changing the status, only emoji and status text, right @nickvergessen?

**Comment by nickvergessen** [comment]
> which is not changing the status, only emoji and status text, right

We do change the status to "Away" (when the user was not invisible (=> no status change) or DND (=> in call with DND))
https://github.com/nextcloud/spreed/blob/3ce5c7bdaafeb8f4747f7bc05e60ebd33ada192a/lib/Status/Listener.php#L55-L83

**Comment by miaulalala** [comment]
> Agree with @nimishavijay. If that is out of scope, a good first step would be to just do the same that we do when in a call – which is not changing the status, only emoji and status text, right @nickvergessen?

Aight, chaningin the status to "away" for now, and removing the icons. Follow up ticket for icon changes discussion here:https://github.com/nextcloud/server/issues/41299

**Review by ChristophWurst — COMMENTED** [review]
Code looks good

**Review by ChristophWurst — APPROVED** [review]
Worked for my local instance once I cleared the *Away* override and the custom status message

File: apps/dav/lib/CalDAV/FreeBusy/FreeBusyGenerator.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ */
+
+namespace OCA\DAV\CalDAV\FreeBusy;
+
+use DateTimeInterface;
+use DateTimeZone;
+use Sabre\VObject\Component\VCalendar;
+
+class FreeBusyGenerator extends \Sabre\VObject\FreeBusyGenerator {
```
[the remark this answers is not in the document]
Comment (miaulalala) [reply apps/dav/lib/CalDAV/FreeBusy/FreeBusyGenerator.php:32]: Not really sure what to do about this here as it's a problem with the lib class

[the remark this answers is not in the document]
Comment (miaulalala) [reply apps/dav/lib/CalDAV/FreeBusy/FreeBusyGenerator.php:32]: Not really sure what to do about this here as it's a problem with the lib class

File: apps/dav/lib/CalDAV/FreeBusy/FreeBusyGenerator.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+	public function __construct(DateTimeInterface $start = null, DateTimeInterface $end = null, $objects = null, DateTimeZone $timeZone = null) {
```
[the remark this answers is not in the document]
Comment (miaulalala) [reply apps/dav/lib/CalDAV/FreeBusy/FreeBusyGenerator.php:34]: Not really sure what to do about this here as it's a problem with the lib class

**Review by AndyScherzinger — APPROVED** [review]

**Comment by miaulalala** [comment]
Failing cypress tests are 

1) Change the login fields then reset them
       Ensure undo button presence:
     AssertionError: Timed out retrying after 4000ms: Expected to find element: `[data-admin-theming-setting-field="name"] .input-field__clear-button`, but never found it.
      at Context.eval (webpack://nextcloud/./cypress/e2e/theming/admin-settings.cy.ts:219:11)


  1) Settings: Create and delete users
       Can create a user with additional field data:
     CypressError: Timed out retrying after 4050ms: `cy.click()` failed because this element:

File: lib/composer/composer/autoload_classmap.php [hunk 3]
```diff
@@ -1203,6 +1203,7 @@
     'OC\\Core\\Migrations\\Version25000Date20220602190540' => $baseDir . '/core/Migrations/Version25000Date20220602190540.php',
     'OC\\Core\\Migrations\\Version25000Date20220905140840' => $baseDir . '/core/Migrations/Version25000Date20220905140840.php',
     'OC\\Core\\Migrations\\Version25000Date20221007010957' => $baseDir . '/core/Migrations/Version25000Date20221007010957.php',
+    'OC\\Core\\Migrations\\Version26000Date20230220105540' => $baseDir . '/core/Migrations/Version26000Date20230220105540.php',
```
Comment (nickvergessen) [inline lib/composer/composer/autoload_classmap.php:1206]: file is missing or autoloader is wrong.

**Comment by miaulalala** [comment]
Failing Cypress test: theming/admin-settings.cy.ts   (unrelated)

**Pull request merged**